### PR TITLE
*: check privilege when reusing the cached plan

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

Once prepared plan cache enabled and the cached plan hit need `privilege check` and `table lock check`

### What is changed and how it works?

- store visitinfo in newly defined `PreparedObject`(try not to add many things in ast structure)
- check privilege when cached plan hit

pull requests #11970  cached execution is faced with same problem(see #11970 comments)

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

Related changes

Release note

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -202,7 +202,7 @@ type SessionVars struct {
 	// systems variables, don't modify it directly, use GetSystemVar/SetSystemVar method.
 	systems map[string]string
 	// PreparedStmts stores prepared statement.
-	PreparedStmts        map[uint32]*ast.Prepared
+	PreparedStmts        map[uint32]interface{}
```
Comment (zz-jason) [inline sessionctx/variable/session.go]: How about directly store `PrepareObject`?

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -79,6 +79,12 @@ type aggHintInfo struct {
 	preferAggToCop bool
 }
 
+// PrepareObject store prepared ast from PrepareExec and other related fields
+type PrepareObject struct {
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: 1. It's better to put this struct definition together with the prepared plan definition.
2. How about s/PrepareObject/CachedPrepareStmt/`?

File: sessionctx/variable/session.go (hunk 1 above)
Comment (cfzjywxk) [reply sessionctx/variable/session.go]: I've tried this. `PrepareObject` stores `visitInfo`(this in planner ) field, seems  `SessionVars` struct in `variable package` cannot  import planner related package, there would be cycle
@lysu  @zz-jason

File: session/session.go [hunk 3]
```diff
@@ -265,16 +265,21 @@ func (s *session) cleanRetryInfo() {
...
+				cacheKey = plannercore.NewPSTMTPlanCacheKey(s.sessionVars, firstStmtID, preparedAst.SchemaVersion)
+			}
+		}
 	}
 	for i, stmtID := range retryInfo.DroppedPreparedStmtIDs {
 		if planCacheEnabled {
-			if i > 0 {
-				plannercore.SetPstmtIDSchemaVersion(cacheKey, stmtID, s.sessionVars.PreparedStmts[stmtID].SchemaVersion)
+			if i > 0 && preparedAst != nil {
+				plannercore.SetPstmtIDSchemaVersion(cacheKey, stmtID, preparedAst.SchemaVersion)
```
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply session/session.go]: This is a problem. Current prepared plan cache logic does not handle this and  rely on lru list length limit to remove unused plan object, maybe we need to do removal in `getPhysicalPlan` when schema version changed

File: server/conn_stmt.go [hunk 4]
```diff
@@ -624,8 +625,12 @@ func (cc *clientConn) handleSetOption(data []byte) (err error) {
 
 func (cc *clientConn) preparedStmt2String(stmtID uint32) string {
 	sv := cc.ctx.GetSessionVars()
-	if prepared, ok := sv.PreparedStmts[stmtID]; ok {
-		return prepared.Stmt.Text() + sv.GetExecuteArgumentsInfo()
+	if preparedPointer, ok := sv.PreparedStmts[stmtID]; ok {
+		preparedObj, ok := preparedPointer.(*plannercore.CachedPrepareStmt)
```
Comment (zz-jason) [inline server/conn_stmt.go]: should we use the form of:
```go
if !ok {
    return errors.Errorf("invalid CachedPrepareStmt type")
}
preparedAst := preparedObj.PreparedAst
...
```

**Comment by cfzjywxk** [comment]
@lysu @zz-jason @jackysp  PTAL thx

**Comment by cfzjywxk** [comment]
@jackysp @lysu PTAL thx

**Comment by cfzjywxk** [comment]
@jackysp @lysu  PTAL thx~

**Comment by jackysp** [comment]
Please resolve the conflict, @cfzjywxk .

**Pull request merged**